app/services/dpul_success_dashboard_report_generator.rb

```diff
@@ -0,0 +1,168 @@
+# frozen_string_literal: true
+require "json"
+
+# frozen_string_literal: true
+# Generates a report of DPUL success metrics given a date range
+# This report is used to measure user engagement with DPUL resources
+class DpulSuccessDashboardReportGenerator
+  attr_reader :date_range
+
+  def initialize(date_range:)
+    @date_range = date_range
+    @totals_hash = { date: "TOTAL", visitors: 0, pageviews: 0, bounce_rate: 0, visit_duration: 0, visits: 0 }
+  end
+
+  def default_dates
+    first_day = @date_range.first.strftime("%Y,%m,%d")
+    last_day = @date_range.last.strftime("%Y,%m,%d")
+    "start: new Date(" + first_day + "), end: new Date(" + last_day + ")"
+  end
+
+  def display_dates
+    @date_range.first.strftime("%B %d, %Y") + " - " + @date_range.last.strftime("%B %d, %Y")
+  end
+
+  def daily_metrics
+    stats_hash = traffic.index_by { |stat| stat["date"] }
+    # Iterate over custom event data hashes and merge data into stats
+    downloads.each do |date_key, metrics|
+      date_str = date_key.to_s
+      stats_hash[date_str]&.merge!(metrics.transform_keys(&:to_s))
+    end
+
+    record_page_views.each do |date_key, metrics|
+      date_str = date_key.to_s
+      stats_hash[date_str]&.merge!(metrics.transform_keys(&:to_s))
+    end
+
+    viewer_clicks.each do |date_key, metrics|
+      date_str = date_key.to_s
+      stats_hash[date_str]&.merge!(metrics.transform_keys(&:to_s))
+    end
+
+    # Convert stats hash back to array
+    stats_with_metrics = stats_hash.values
+    stats_with_metrics << @totals_hash.transform_keys(&:to_s)
+  end
+
+  def traffic_request
```

Comment:
Non-blocking, but I wonder if we can dedup some of this boilerplate with something like

```ruby
def plausible_request
    Faraday.new(url: "https://plausible.io") do |conn|
      conn.request :authorization, "Bearer", Figgy.config["plausible_api_key"]
      conn.adapter Faraday.default_adapter
      conn.headers["Content-Type"] = "application/json"
      conn.params["site_id"] = "dpul.princeton.edu"
      yield conn
    end
end
```

Reply:
Done!